src/main/java/com/aws/greengrass/lifecyclemanager/Lifecycle.java

```diff
@@ -629,22 +630,27 @@
 
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
     private void handleCurrentStateStopping() throws InterruptedException {
-        // does not handle desiredState in STOPPING because we must stop first.
-        // does not use setBackingTask because it will cancel the existing task.
-        Future<?> shutdownFuture = greengrassService.getContext().get(ExecutorService.class).submit(() -> {
-            try {
-                greengrassService.shutdown();
-            } catch (InterruptedException i) {
-                logger.atWarn("service-shutdown-interrupted").log("Service interrupted while running shutdown");
-            } catch (Throwable t) {
-                greengrassService.serviceErrored(t);
-            }
-        });
-
+        Future<?> shutdownFuture = null;
         try {
-            Integer timeout = getTimeoutConfigValue(
+            int timeout = getTimeoutConfigValue(
                         LIFECYCLE_SHUTDOWN_NAMESPACE_TOPIC, DEFAULT_SHUTDOWN_STAGE_TIMEOUT_IN_SEC);
-            shutdownFuture.get(timeout, TimeUnit.SECONDS);
+            Instant waitStart = Instant.now();
+            waitForBackingTask(Duration.ofSeconds(timeout));
+            long timeWaited = ChronoUnit.SECONDS.between(waitStart, Instant.now());
+
+            // does not handle desiredState in STOPPING because we must stop first.
+            // does not use setBackingTask because it will cancel the existing task.
+            shutdownFuture = greengrassService.getContext().get(ExecutorService.class).submit(() -> {
+                try {
+                    greengrassService.shutdown();
+                } catch (InterruptedException i) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

**Remember**: Do this only for InterruptedException. Doing this for other types of exceptions may give rise to subtle bugs.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)